Login rework

tadeboro opened the pull request:
This pull request introduces semi-automatic login session renewal. Unfortunately, we were forced to change quite a bit of code since login functionality was moved one abstraction layer down in the library: from the resource level to connector level.

Fixes #12

tadeboro commented on lib/redfish_client/connector.rb:
```diff
@@ -52,10 +68,18 @@ def remove_headers(headers)
+
+      @connection.get(path: path, headers: @headers).tap do |r|
+        @cache[path] = r if r.status == 200
+      end
```
Above three lines are equivalent to:

    r = @connection.get(path: path, headers: @headers)
    if r.status == 200
      @cache[path] = r
    end
    r

`tap` always returns `self`, block is used for side-effects only, which makes it really useful in situations such as above, where we can remove the temporary variable and unclutter the code.

matejart commented on lib/redfish_client/resource.rb:
```diff
@@ -103,13 +103,6 @@ def respond_to_missing?(symbol, include_private = false)
-    def reset
-      @connector.reset if @connector.respond_to?(:reset)
-    end
-
```
Lines 18-21 talk about the `reset` method, Is that still relevant?

matejart replied:
If you remove these methods, is there any other way to manage cache? From what I've seen, the functionality is now only in the connector, which is not exposed to the developers. What is the approach for requesting a cache invalidation and getting a new value from the server?

tadeboro replied:
I was wondering why this was not as obvious as I thought it was. And then I found out that I did not create a separate commit that introduced this functionality. So I did what I should have done when initially committing changes: split this into separate commit. Thanks for spotting this.

matejart commented on lib/redfish_client/connector.rb:
```diff
@@ -73,11 +79,11 @@ def remove_headers(headers)
     #
     # @param path [String] path to the resource, relative to the base url
-    # @return [Excon::Response] response object
+    # @return [Response] response object
```
Would it make sense to use `Connector::Response` here as well to be consistent with the documentaiton elsewhere? Or is it a convention to omit what is actually the defining class?

tadeboro replied:
Documentation generator will this resolve properly, so there is no need to prefix the type with parent module/class.

matejart commented on spec/redfish_client/resource_spec.rb:
```diff
@@ -269,7 +269,7 @@
   context "#delete" do
     it "returns response instance" do
-      expect(resource.delete).to be_a Excon::Response
+      expect(resource.delete).to be_a RedfishClient::Connector::Response
```
Nitpicking: you added parentheses `()` in the other test cases, but not here. In an earlier commit you removed them elsewhere.
In my opinion it is fine without them.

tadeboro replied:
It seems I missed that one. Usually, the only method call that I leave without parentheses is `.to`. This adds a nice visual separation of test and expected result. Since this test was changed later on, I will leave this as is.

matejart commented on spec/redfish_client_spec.rb:
```diff
@@ -5,63 +5,43 @@
-      expect(client.sub.key).to eq("custom_val")
+      stub_request(:get, "http://example.com/redfish/v1")
+        .to_return(status: 200, body: '{"e": "f"}')
+        .to_raise("Should not be here")
```
I really like this solution 👍

matejart commented on lib/redfish_client/connector.rb:
```diff
@@ -126,10 +133,57 @@ def reset(path = nil)
+
+    # Authenticate against the service.
+    #
+    # Calling this method will try to uthenticate against API using
```
**a**uthenticate

tadeboro commented on spec/redfish_client/connector_spec.rb:
```diff
@@ -178,4 +190,106 @@
+      connector.set_auth_info("user", "pass", "/test")
+      connector.login
+
+      connector.logout
```
I was also thinking about this when I wrote this test. Ultimately I decided to skip this since testing re-authentication and other stuff is really not the purpose of this test (we already do this elsewhere). This test asserts that if we need to invalidate credentials when using basic authentication, no API calls are made and things do not blow up.

Anything else would be better suited for integration testing.

tadeboro commented on spec/redfish_client/connector_spec.rb:
```diff
@@ -178,4 +190,106 @@
+      connector.set_auth_info("user", "pass", "/test", "/sessions")
+      connector.login
+
+      connector.logout
```
Not sure why I wrote test without asserts on deletes, since this is the main thing we are testing. I will add an assert statement at the end of the test.

tadeboro commented on spec/redfish_client/root_spec.rb:
```diff
@@ -6,6 +6,41 @@
+        "a" => { "@odata.id" => "/t" },
+        "Links" => { "Sessions" => { "@odata.id" => "/s" } },
+      }
+      described_class.new(connector, raw: raw).login("u", "p")
```
Again, this is just me being silly and not adding any asserts.

tadeboro commented:
Thanks for the comments @matejart. I fixed all the stings that I messed up and things should be in better shape now. Main things I did:

 1. split last commit into three separate commints with clearly stated intentions,
 2. fix the useless logout tests,
 3. touch up the spelling  in comments and
 4. update library documentation (remove references to methods that do not exist anymore).
